### PR TITLE
Feature: 2FA Recovery Codes

### PR DESCRIPTION
**Pull request by juniper-wright**
**Previously:**

* Users were able to provide a second authorization factor either by using Google Authenticator (or similar), or by receiving a 2FA code via SMS

**Now:**

* In addition to GA or SMS, users can now be provided with a list of recovery codes, each of which are one-time-use codes that provide a second authentication factor.

**Notes:**

* Tests have now been added for 2FA recovery codes

### CODE REVIEW

**Review by gimenete — COMMENTED** [review]
I thought we were going to store the backup codes in a column in `auth_users` as an array. Instead of creating a new table.

**Comment by juniper-wright** [comment]
@gimenete I tried that, but it got _very_ messy due to the differences in database limitations. MySQL, for instance, has no arrays, so the only way I could've made it work would've been a comma-delimited string. This seemed infinitely more elegant.

File: src/database/mongodb.js [hunk 1]
```diff
@@ -35,9 +39,33 @@ module.exports = env => {
     }),
     findUserByProviderLogin: sanitizeOutputs(function(login) {
       return db.collection('auth_providers').findOne({ login }).then(function (provider) {
+        if(!provider) {
```
Comment (gimenete) [inline src/database/mongodb.js]: Coding style, whitespace missing.

File: src/index.js [hunk 2]
```diff
@@ -120,7 +120,7 @@ exports.createRouter = (config = {}) => {
 
   const router = express.Router()
   router.use(bodyParser.urlencoded({ extended: false }))
-  router.use(recaptcha.middleware())
+  //router.use(recaptcha.middleware())
```
Comment (gimenete) [inline src/index.js]: This looks like a temporary change but it's still there.

File: src/index.js [hunk 3]
```diff
@@ -227,6 +236,10 @@ exports.createRouter = (config = {}) => {
       .catch(next)
   })
 
+  router.get('/landing', authenticated, (req, res, next) => {
```
Comment (gimenete) [inline src/index.js]: This should only be there for testing.

File: src/index.js [hunk 4]
```diff
@@ -349,7 +363,8 @@ exports.createRouter = (config = {}) => {
     const url = speakeasy.otpauthURL({
       secret: twofactorSecret,
       encoding: 'base32',
-      label: encodeURIComponent(labelForQr(user))
+      label: user.email,
+      issuer: projectName
```
Comment (gimenete) [inline src/index.js:368]: 👍

File: src/index.js [hunk 5]
```diff
@@ -534,7 +578,15 @@ exports.createRouter = (config = {}) => {
           return redirect(user)
             .then(url => res.redirect(url))
         } else {
-          return Promise.reject(new Error('INVALID_TOKEN'))
+          return users.useRecoveryCode(user.id, token)
+            .then(recoveryValidates => {
+              if(recoveryValidates) {
```
Comment (gimenete) [inline src/index.js]: Coding style. Whitespace missing.

File: src/locales/en.json [hunk 6]
```diff
@@ -10,7 +10,7 @@
   "EMAIL_CONFIRMATION_SENT": "Before signing in, please confirm your email address. Check your inbox and follow the instructions",
   "INTERNAL_ERROR": "Internal error",
   "UNKNOWN_ERROR": "Unknown error",
-  "EMAIL_CONFIRMED": "Thank you! Your email address is not confirmed. Please sign in",
+  "EMAIL_CONFIRMED": "Thank you! Your email address is now confirmed. Please sign in",
```
Comment (gimenete) [inline src/locales/en.json:13]: 👍

**Comment by gimenete** [comment]
I made some comments. Also, one question, are we deleting the backup codes when the user disables 2FA?

**Comment by juniper-wright** [comment]
We are now: https://github.com/clevertech/authentication-service/pull/18/commits/2837f59d7a061f6716b0ab04d83fa73e0808c98f!

**Pull request merged**